hazelcast/src/test/java/com/hazelcast/spi/discovery/impl/DiscoveryServicePropertiesBuilderTest.java

```diff
@@ -0,0 +1,128 @@
+import com.hazelcast.config.properties.ValueValidator;
+import com.hazelcast.test.HazelcastParallelClassRunner;
+import com.hazelcast.test.annotation.ParallelTest;
+import com.hazelcast.test.annotation.QuickTest;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+
+import java.util.Collection;
+import java.util.Map;
+
+import static com.hazelcast.config.properties.PropertyTypeConverter.STRING;
+import static java.util.Collections.emptyList;
+import static java.util.Collections.emptyMap;
+import static java.util.Collections.singletonList;
+import static java.util.Collections.singletonMap;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+
+@RunWith(HazelcastParallelClassRunner.class)
+@Category({QuickTest.class, ParallelTest.class})
+public class DiscoveryServicePropertiesBuilderTest {
+    private static final String PROPERTY_KEY_1 = "property1";
+    private static final String PROPERTY_VALUE_1 = "propertyValue1";
+    private static final PropertyDefinition PROPERTY_DEFINITION_1 = new SimplePropertyDefinition(PROPERTY_KEY_1, STRING);
+
+    private DiscoveryServicePropertiesBuilder propertiesBuilder = new DiscoveryServicePropertiesBuilder();
+
+    @Test
+    public void correctProperties() {
+        // given
+        Map<String, Comparable> properties = singletonMap(PROPERTY_KEY_1, (Comparable) PROPERTY_VALUE_1);
+        Collection<PropertyDefinition> propertyDefinitions = singletonList(PROPERTY_DEFINITION_1);
+
+        // when
+        Map<String, Comparable> result = propertiesBuilder.buildProperties(properties, propertyDefinitions);
+
+        // then
+        assertEquals(PROPERTY_VALUE_1, result.get(PROPERTY_KEY_1));
+    }
+
+    @Test(expected = InvalidConfigurationException.class)
+    public void unsatisfiedRequiredProperty() {
+        // given
+        Map<String, Comparable> properties = emptyMap();
+        Collection<PropertyDefinition> propertyDefinitions = singletonList(PROPERTY_DEFINITION_1);
+
+        // when
+        propertiesBuilder.buildProperties(properties, propertyDefinitions);
+
+        // then
+        // throw exception
+    }
+
+    @Test
+    public void unsatisfiedOptionalProperty() {
+        // given
+        Map<String, Comparable> properties = emptyMap();
+        Collection<PropertyDefinition> propertyDefinitions = singletonList(
+                (PropertyDefinition) new SimplePropertyDefinition(PROPERTY_KEY_1, true, STRING));
+
+        // when
+        Map<String, Comparable> result = propertiesBuilder.buildProperties(properties, propertyDefinitions);
+
+        // then
+        assertTrue(result.isEmpty());
+    }
+
+    @Test(expected = ValidationException.class)
+    public void invalidProperty() {
+        // given
+        Map<String, Comparable> properties = singletonMap(PROPERTY_KEY_1, (Comparable) PROPERTY_VALUE_1);
+        Collection<PropertyDefinition> propertyDefinitions = singletonList(
```

Comment:
flavor warning: for such trivial dependencies like the DummyValidator I normally use a mock. F'DummyValidator' doesn't explain its purpose (will it accept or reject? With a mock it becomes more obvious from the test since you record the expected behavior.

I'm a very cautious mock user; I have seen enough code bases with intense damage due to over-mocking. But this is  good example of where a mock would be a perfect fit.

Reply:
Ok, changed it.